src/edit_distance/osa.gleam

```diff
@@ -0,0 +1,202 @@
+  i: Int,
+  j: Int,
+) -> Int {
+  let cost = match_cost(one, i - 1, other, j - 1)
+  // Calculate deletion cost
+  let d_i1j = get_cost(d, i - 1, j) + 1
+  // Calculate insertion cost
+  let d_ij1 = get_cost(d, i, j - 1) + 1
+  // Calculate substitution cost
+  let d_i1j1 = get_cost(d, i - 1, j - 1) + cost
+
+  // Find the minimum cost between insertion, deletion, and substitution
+  let min_cost = min_three(d_i1j, d_ij1, d_i1j1)
+
+  // Finally, take into account the cost of a transposition
+  let updated_cost = case i > 1 && j > 1 {
+    True -> {
+      case check_transposition(one, i - 1, other, j - 1) {
+        True -> int.min(min_cost, get_cost(d, i - 2, j - 2) + 1)
+        False -> min_cost
+      }
+    }
+    False -> min_cost
+  }
+  updated_cost
+}
+
+// A function that retrieves an edit cost from the dictionary representing a distance matrix.
+// The function retrieves the cost using the tuple (row_index, column_index) as the key. 
+// The value corresponds to the distance calculated between the substrings: 
+// - 'one' (from position 0 to row_index)
+// - 'other' (from position 0 to column_index) 
+// Since the matrix (and thus the dictionary) is processed iteratively (row-wise and column-wise), 
+// the value at index (row_index, column_index), where row_index < i and column_index < j, 
+// will already be present in the matrix by the time we calculate the value for entry (i, j). We 
+// can thus access this entry with confidence
+fn get_cost(d: Dict(#(Int, Int), Int), row: Int, column: Int) -> Int {
+  let assert Ok(cost) = dict.get(d, #(row, column))
+  cost
+}
+
+// Check if a transposition is valid by comparing adjacent characters in both strings
+fn check_transposition(
+  one: Dict(Int, String),
+  i: Int,
+  other: Dict(Int, String),
+  j: Int,
+) -> Bool {
+  let one_i = dict.get(one, i)
+  let other_j = dict.get(other, j)
+  let one_i1 = dict.get(one, i - 1)
+  let other_j1 = dict.get(other, j - 1)
+
+  // Check if transposing adjacent characters results in a match
+  case one_i, other_j1, one_i1, other_j {
+    Ok(c0), Ok(c1), Ok(c2), Ok(c3) if c0 == c1 && c2 == c3 -> True
+    _, _, _, _ -> False
+  }
+}
+
+// Compute the edit distance by filling the distance matrix with the minimum costs
+fn compute_distances(
+  one: Dict(Int, String),
+  other: Dict(Int, String),
+  d: Dict(#(Int, Int), Int),
+  m: Int,
+  n: Int,
+) -> Dict(#(Int, Int), Int) {
+  // Initialize base cases, to avoid continuously checking these later on
+  // Fixing i = 0 or j = 0, set the value of transforming one string into the other
+  let d = {
+    use acc_d, i <- list.fold(list.range(0, m), d)
+    dict.insert(acc_d, #(i, 0), i)
+  }
+  let d = {
+    use acc_d, j <- list.fold(list.range(0, n), d)
+    dict.insert(acc_d, #(0, j), j)
+  }
+
+  // Iterate over the matrix to calculate the minimum costs for all substrings
+  use acc_d, i <- list.fold(list.range(1, m), d)
+  use inner_d, j <- list.fold(list.range(1, n), acc_d)
```

Comment:
I'd like this to be as efficient as possible, calling `list.range` creates an entire list in memory to fold over it. Here you could define your own `fold_range` function that looks like this:
```gleam
fn fold_range(from start: Int, to end: Int, acc acc: a, do fun: fn(a, Int) -> a) -> a {
  // This calls recursively from start to end without
  // having to build a possibly big list in memory
}
```
And then this code would become:
```gleam
use acc, i <- fold_range(from: 1, to: m, acc: d)
use acc, j <- fold_range(from: 1, to: n, acc: acc)
...
```